py/bitbox02/bitbox02/bitbox02/bitbox02.py

```diff
@@ -423,11 +430,47 @@ def btc_sign(
                         script_config_index=tx_input["script_config_index"],
                     )
                 )
+                if supports_antiklepto and is_inputs_pass2:
+                    host_nonce = os.urandom(32)
+                    request.btc_sign_input.host_nonce_commitment.commitment = antiklepto_host_commit(
+                        host_nonce
+                    )
+
                 next_response = self._msg_query(
                     request, expected_response="btc_sign_next"
                 ).btc_sign_next
-                if next_response.has_signature:
+
+                if supports_antiklepto and is_inputs_pass2:
+                    assert next_response.type == btc.BTCSignNextResponse.HOST_NONCE
+                    assert next_response.HasField("anti_klepto_signer_commitment")
+                    signer_commitment = next_response.anti_klepto_signer_commitment.commitment
+
+                    btc_request = btc.BTCRequest()
+                    btc_request.antiklepto_signature.CopyFrom(
+                        antiklepto.AntiKleptoSignatureRequest(host_nonce=host_nonce)
+                    )
+                    next_response = self._btc_msg_query(
+                        btc_request, expected_response="sign_next"
+                    ).sign_next
+
+                    if self.debug:
+                        print(
+                            f"For input {input_index}, the host contributed the nonce {host_nonce.hex()}"
+                        )
+
+                    assert next_response.has_signature
+                    antiklepto_verify(host_nonce, signer_commitment, next_response.signature)
+
+                    if self.debug:
+                        print(f"Antiklepto nonce verification PASSED for input {input_index}")
+
+                if is_inputs_pass2:
+                    assert next_response.has_signature
                     sigs.append((input_index, next_response.signature))
+
+                if input_index == len(inputs) - 1:
+                    is_inputs_pass2 = True
```

Comment:
This is not so nice, as it is mixing two styles of API:
- explicit sequence of requests
- implicit sequence: respond with whatever the device asks

I am considering getting rid of the Next-driven API, and make the sequence completely explicit.

Reply:
+1 for explicit

Reply:
:+1: will rewrite it at a later point to switch away from the next-driven style. not in this PR though.